rts/motoko-rts/src/gc/mark_compact/mark_stack.rs

```diff
@@ -50,20 +50,25 @@ unsafe fn grow_stack<M: Memory>(mem: &mut M) {
     (*STACK_BLOB_PTR).len = new_cap.to_bytes();
 }
 
-pub unsafe fn push_mark_stack<M: Memory>(mem: &mut M, obj: usize) {
+pub unsafe fn push_mark_stack<M: Memory>(mem: &mut M, obj: usize, obj_tag: Tag) {
+    // We add 2 words in a push, and `STACK_PTR` and `STACK_TOP` are both multiples of 2, so we can
+    // do simple equality check here
     if STACK_PTR == STACK_TOP {
         grow_stack(mem);
     }
 
     *STACK_PTR = obj;
-    STACK_PTR = STACK_PTR.add(1);
+    *(STACK_PTR.add(1)) = obj_tag as usize;
+    STACK_PTR = STACK_PTR.add(2);
 }
 
-pub unsafe fn pop_mark_stack() -> Option<usize> {
+pub unsafe fn pop_mark_stack() -> Option<(usize, Tag)> {
```

Comment:
I guess you could return a sentinal, not an option, for perf, but perhaps not worth the effort.

Reply:
Maybe Rust even does that automatically? It does for some types, but the list doens't mention tuples :https://doc.rust-lang.org/std/option/#representation

Reply:
It seems it does! https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5007ac15b58e9dca74e4b54789e139f2